File: C++/Medium/Accepted/course-schedule-iv_7.cpp
class Solution {
public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        
        vector<vector<int>> graph(numCourses, vector<int>(numCourses, 0));
        for (auto& e: prerequisites) {
            graph[e[0]][e[1]] = 1;
        }
        for (int i = 0; i < numCourses; ++i) {
            graph[i][i] = 1;
        }
        for (int k = 0; k < numCourses; ++k) {
            for (int i = 0; i < numCourses; ++i) {
                for (int j = 0; j < numCourses; ++j) {
                    graph[i][j] = graph[i][j] || (graph[i][k] && graph[k][j]);
                }
            }
        }
        vector<bool> ans;
        for (auto& q: queries) {
            ans.push_back(graph[q[0]][q[1]]);
        }
        return ans;
    }
};

File: Rust/Medium/CompileError/course-schedule-iv_7.rs
impl Solution {
    pub fn check_if_prerequisite(num_courses: i32, prerequisites: Vec<Vec<i32>>, queries: Vec<Vec<i32>>) -> Vec<bool> {
        
        let mut graph = vec![vec![]; num_courses as usize];
        let mut dp = vec![vec![false; num_courses as usize]; num_courses as usize];
        for (i, j) in prerequisites {
            graph[i as usize].push(j as usize);
        }
        for i in 0..num_courses as usize {
            let mut queue = vec![i];
            while let Some(j) = queue.pop() {
                for &k in &graph[j] {
                    if !dp[i][k] {
                        dp[i][k] = true;
                        queue.push(k);
                    }
                }
            }
        }
        queries.iter().map(|q| dp[q[0] as usize][q[1] as usize]).collect()
    }
}